pkg/resources/storage_integration.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	if data.HasChange("enabled") {
+		runSetStatement = true
+		stmt.SetBool(`ENABLED`, data.Get("enabled").(bool))
+	}
+
+	if data.HasChange("storage_allowed_locations") {
+		runSetStatement = true
+		stmt.SetStringList("STORAGE_ALLOWED_LOCATIONS", expandStringList(data.Get("storage_allowed_locations").([]interface{})))
+	}
+
+	// We need to UNSET this if we remove all storage blocked locations. I don't think
+	// this is documented by Snowflake, but this is how it works.
+	//
+	// @TODO move the SQL back to the snowflake package
+	if data.HasChange("storage_blocked_locations") {
+		v := data.Get("storage_blocked_locations").([]interface{})
+		if len(v) == 0 {
+			err := DBExec(db, fmt.Sprintf(`ALTER STORAGE INTEGRATION %v UNSET STORAGE_BLOCKED_LOCATIONS`, data.Id()))
+			if err != nil {
+				return fmt.Errorf("error unsetting storage_blocked_locations: %w", err)
+			}
+		} else {
+			runSetStatement = true
+			stmt.SetStringList("STORAGE_BLOCKED_LOCATIONS", expandStringList(v))
+		}
+	}
+
+	if data.HasChange("storage_provider") {
+		runSetStatement = true
+		setStorageProviderSettings(data, stmt)
+	} else {
+		if data.HasChange("storage_aws_role_arn") {
+			runSetStatement = true
+			stmt.SetString("STORAGE_AWS_ROLE_ARN", data.Get("storage_aws_role_arn").(string))
+		}
+		if data.HasChange("azure_tenant_id") {
+			runSetStatement = true
+			stmt.SetString("AZURE_TENANT_ID", data.Get("azure_tenant_id").(string))
+		}
+	}
+
+	if runSetStatement {
+		if err := DBExec(db, stmt.Statement()); err != nil {
+			return fmt.Errorf("error updating storage integration: %w", err)
+		}
+	}
+
+	return ReadStorageIntegration(data, meta)
+}
+
+// DeleteStorageIntegration implements schema.DeleteFunc
+func DeleteStorageIntegration(data *schema.ResourceData, meta interface{}) error {
+	return DeleteResource("", snowflake.StorageIntegration)(data, meta)
+}
+
+// StorageIntegrationExists implements schema.ExistsFunc
+func StorageIntegrationExists(data *schema.ResourceData, meta interface{}) (bool, error) {
+	db := meta.(*sql.DB)
+	id := data.Id()
+
+	stmt := snowflake.StorageIntegration(id).Show()
+	rows, err := db.Query(stmt)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		return true, nil
+	}
+	return false, nil
+}
+
+func setStorageProviderSettings(data *schema.ResourceData, stmt snowflake.SettingBuilder) error {
+	storageProvider := data.Get("storage_provider").(string)
+	stmt.SetString("STORAGE_PROVIDER", storageProvider)
+
+	switch storageProvider {
+	case "S3":
+		v, ok := data.GetOk("storage_aws_role_arn")
+		if !ok {
+			return fmt.Errorf("If you use the S3 storage provider you must specify a storage_aws_role_arn")
+		}
+		stmt.SetString(`STORAGE_AWS_ROLE_ARN`, v.(string))
+	case "AZURE":
+		v, ok := data.GetOk("azure_tenant_id")
+		if !ok {
+			return fmt.Errorf("If you use the Azure storage provider you must specify an azure_tenant_id")
+		}
+		stmt.SetString(`AZURE_TENANT_ID`, v.(string))
+	}
+
```

Comment:
Looks like you need to add a case for `GCS` if you're allowing it here: https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/140/files#diff-ea99148ac4d4624cd63de6823b620da4R52

Would also be good to add a default statement so that it errors if there are any other providers given. Same goes for the other switch case statements.

Reply:
There is nothing to set for GCS so I have added a no op case.
I also added a default, but note that the provider validates the storage provider so it shouldn't be possible for this to happen.